src/NuGet.Core/NuGet.PackageManagement/PackagePreFetcher.cs

```diff
@@ -0,0 +1,206 @@
+        /// </summary>
+        public static async Task<Dictionary<PackageIdentity, PackagePreFetcherResult>> GetPackagesAsync(
+            IEnumerable<NuGetProjectAction> actions,
+            FolderNuGetProject packagesFolder,
+            Configuration.ISettings settings,
+            Common.ILogger logger,
+            CancellationToken token)
+        {
+            if (token == null)
+            {
+                throw new ArgumentNullException(nameof(token));
+            }
+
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            if (packagesFolder == null)
+            {
+                throw new ArgumentNullException(nameof(packagesFolder));
+            }
+
+            if (actions == null)
+            {
+                throw new ArgumentNullException(nameof(actions));
+            }
+
+            var result = new Dictionary<PackageIdentity, PackagePreFetcherResult>();
+            var maxParallelTasks = PackageManagementConstants.DefaultMaxDegreeOfParallelism;
+            var toDownload = new Queue<NuGetProjectAction>();
+            var seen = new HashSet<PackageIdentity>();
+
+            // Find all uninstalled packages
+            var uninstalledPackages = new HashSet<PackageIdentity>(
+                actions.Where(action => action.NuGetProjectActionType == NuGetProjectActionType.Uninstall)
+                .Select(action => action.PackageIdentity));
+
+            // Check the packages folder for each package
+            // If the package is not found mark it for download
+            // These actions need to stay in order!
+            foreach (var action in actions)
+            {
+                // Ignore uninstalls here
+                if (action.NuGetProjectActionType == NuGetProjectActionType.Install
+                    && !seen.Contains(action.PackageIdentity))
+                {
+                    // Avoid duplicate downloads
+                    seen.Add(action.PackageIdentity);
+
+                    string installPath = null;
+
+                    // Packages that are also being uninstalled cannot come from the
+                    // packages folder since it will be gone. This is true for reinstalls.
+                    if (!uninstalledPackages.Contains(action.PackageIdentity))
+                    {
+                        // Check the packages folder for the id and version
+                        installPath = packagesFolder.GetInstalledPackageFilePath(action.PackageIdentity);
+
+                        // Verify the nupkg exists
+                        if (!File.Exists(installPath))
+                        {
+                            installPath = null;
+                        }
+                    }
+
+                    if (!string.IsNullOrEmpty(installPath))
```

Comment:
it will contain the full path to the already installed package if it exists, otherwise it will be empty, which means that the package should be downloaded